core/templates/dev/head/domain/learner_dashboard/LearnerPlaylistService.js

```diff
@@ -0,0 +1,117 @@
+            successfullyAdded = false;
+            alertsService.addInfoMessage(
+              'You have already completed or are completing this ' +
+              'activity.');
+          }
+          if (response.data.belongs_to_subscribed_activities) {
+            successfullyAdded = false;
+            alertsService.addInfoMessage(
+              'This is present in your creator dashboard');
+          }
+          if (response.data.playlist_limit_exceeded) {
+            successfullyAdded = false;
+            alertsService.addInfoMessage(
+              'Your \'Play Later\' list is full!  Either you can ' +
+              'complete some or you can head to the learner dashboard ' +
+              'and remove some.');
+          }
+          if (successfullyAdded) {
+            alertsService.addSuccessMessage(
+              'Successfully added to your \'Play Later\' list.');
+          }
+        });
+      return successfullyAdded;
+    };
+
+    var _removeFromLearnerPlaylist = function(activityId, activityTitle,
+      activityType, learnerDashboardActivityIds) {
+      $modal.open({
+        templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+          '/pages/learner_dashboard/' +
+          'remove_activity_from_learner_dashboard_modal_directive.html'),
+        backdrop: true,
+        resolve: {
+          activityId: function() {
+            return activityId;
+          },
+          activityTitle: function() {
+            return activityTitle;
+          }
+        },
+        controller: [
+          '$scope', '$modalInstance', '$http',
+          'UrlInterpolationService', function($scope, $modalInstance,
+            $http, UrlInterpolationService) {
+            $scope.sectionNameI18nId = (
+              'I18N_LEARNER_DASHBOARD_PLAYLIST_SECTION');
+            $scope.activityTitle = activityTitle;
+            var removeFromLearnerPlaylistUrl = (
+              UrlInterpolationService.interpolateUrl(
+                '/learnerplaylistactivityhandler/' +
+                '<activityType>/<activityId>', {
+                  activityType: activityType,
+                  activityId: activityId
+                }));
+            $scope.remove = function() {
+              $http['delete'](removeFromLearnerPlaylistUrl);
+              $modalInstance.close();
+            };
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+            };
+          }
+        ]
+      }).result.then(function() {
+        if (activityType === constants.ACTIVITY_TYPE_EXPLORATION) {
+          learnerDashboardActivityIds.removeFromExplorationLearnerPlaylist(
+            activityId);
+        } else if (activityType === constants.ACTIVITY_TYPE_COLLECTION) {
+          learnerDashboardActivityIds.removeFromCollectionLearnerPlaylist(
+            activityId);
+        }
+      });
+    };
+
+    var _fetchLearnerDashboardData = function() {
```

Comment:
Shouldn't this service be maintaining a private copy of the activity ids object? You should only be doing a single fetch.

Or rather: the public API for this service should be something like fetchLearnerDashboardActivityIds(), and the logic in that should be "if data exists, return promise containing data, else make new fetch from backend" or similar.

Reply:
Ohh I think this is included in this service by mistake. This is actually the service for fetching the data of the learner dashboard in which case this is correct. I have removed this from here. Sorry for this!